src/rvsdg/optimize_direct_jumps.rs

```diff
@@ -0,0 +1,188 @@
+                    .edges_directed(previous, Direction::Outgoing)
+                    .collect::<Vec<_>>();
+                // single outgoing edge from previous
+                // and two distinct nodes
+                if previous_outgoing.len() == 1 && previous != node {
+                    let previous_node = node_mapping[&previous];
+
+                    // this node will be mapped to the previous
+                    node_mapping.insert(node, previous_node);
+
+                    // add instructions to the end of the previous node
+                    resulting_graph[previous_node]
+                        .instrs
+                        .extend(self.simple_func.graph[node].instrs.to_vec());
+
+                    collapse_node = true;
+                }
+            }
+
+            if !collapse_node {
+                // add the node
+                let new_node = resulting_graph.add_node(self.simple_func.graph[node].clone());
+                node_mapping.insert(node, new_node);
+
+                edges_to_add.extend(
+                    self.simple_func
+                        .graph
+                        .edges_directed(node, Direction::Incoming),
+                );
+            }
+        }
+
+        for edge in edges_to_add {
+            let source = node_mapping[&edge.source()];
+            let target = node_mapping[&edge.target()];
+            resulting_graph.add_edge(source, target, edge.weight().clone());
+        }
+
+        SimpleCfgFunction {
+            name: self.simple_func.name.clone(),
+            args: self.simple_func.args.clone(),
+            graph: resulting_graph,
+            entry: node_mapping[&self.simple_func.entry],
+            exit: node_mapping[&self.simple_func.exit],
+            _phantom: Simple,
+            return_ty: self.simple_func.return_ty.clone(),
+        }
+    }
+}
+
+impl SimpleCfgFunction {
+    pub fn optimize_jumps(&self) -> Self {
+        JumpOptimizer { simple_func: self }.optimized()
+    }
+}
+
+impl SimpleCfgProgram {
+    pub fn optimize_jumps(&self) -> Self {
+        SimpleCfgProgram {
+            functions: self.functions.iter().map(|f| f.optimize_jumps()).collect(),
+        }
+    }
+}
+
+#[test]
+fn single_node() {
+    // TODO these imports are very bad
+    use super::super::cfg::tests::cfg_test_equiv;
+    use crate::cfg::BlockName;
+
+    let mut graph = StableDiGraph::new();
+    let node = graph.add_node(BasicBlock {
+        name: BlockName::Entry,
+        instrs: vec![],
+        footer: vec![],
+        pos: None,
+    });
+    let input_cfg = SimpleCfgFunction {
+        args: vec![],
+        graph,
+        entry: node,
+        exit: node,
+        name: "test".to_string(),
+        _phantom: Simple,
+        return_ty: None,
+    };
+
+    cfg_test_equiv!(input_cfg.optimize_jumps(), []);
+}
+
+#[test]
+fn loops_to_self() {
+    // TODO you have to import both of these
+    //  is there a way to package up some macros all together?
```

Comment:
You should be able to  put these macros in their own file and then use them in all these different test files. Is there a reason that wasn't working? (either way, I think you can do an absolute path rather than `super::super`.

Reply:
fixed!